Fixes for various bugs introduced by stdenv update in staging

ttuegel opened the pull request:
### Motivation

`staging` is broken.

### Testing

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

The bugfixes in this branch work, but I am not done testing to check that these are *all* the blocking bugs in `staging`.

Mic92 committed:
alacritty: init at 2017-07-08

fixes #21801

aherrmann committed:
llvm_4: Fix debugVersion

Previously failed on installation phase
when `substituteInPlace` could not find the cmake exports file.

mbrock committed:
dapp: init at v0.5.2

Simple tool for creating Ethereum-based dapps

florianjacob committed:
journalwatch & journalwatch service: init at 1.1.0

jtojnar committed:
gst_all_1.gst-plugins-bad: allow building gtksink plugin

Corebird requires gtksink gstreamer plugin to play videos. [1] The plugin,
however, is only built when GTK is available.

This patch adds gtk3 as an optional dependency to gst_all_1.gst-plugins-bad
package, allowing the build of gtksink.

[1]: https://github.com/baedert/corebird/issues/431

Nadrieril committed:
bitlbee: Add optional libpurple dependency

Nadrieril committed:
bitlbee service: Add option to load libpurple plugins into bitlbee

peti committed:
haskell-hsyslog: update version override

And move it from configuration-nix to configuration-common.

nbp committed:
Update Mozilla Software Releases sub-key.

This commit update the Mozilla Software Releases, with the key found at:
https://gpg.mozilla.org/pks/lookup?search=Mozilla+Software+Releases+%3Crelease%40mozilla.com%3E&op=vindex

calvertvl committed:
calibre: fix html5lib error

peti committed:
haskell-hledger-iadd: drop obsolete override (the build fails with or without it)

mdorman committed:
elpa-packages: 2017-07-17

mdorman committed:
org-packages: 2017-07-17

mdorman committed:
melpa-stable-packages: 2017-07-17

mdorman committed:
melpa-packages: 2017-07-17

ttuegel committed:
breeze-icons: generate icon cache

Fixes #21345 by generating the icon cache when the theme is installed.

np committed:
Keybase updates

keybase: 1.0.20 -> 1.0.22
keybase-gui: 1.0.23-20170519175207.d6c5e9e -> 1.0.25-20170714172717.73f9070
kbfs: 20170209.d1db463 -> 20170429.44efa06

teh committed:
python.pkgs.pytorch: init at 0.1.12

jfrankenau committed:
solvespace: Fix path in desktop entry file

FRidh committed:
Merge pull request #27290 from calvertvl/fix_calibre_html5lib

calibre: fix html5lib error

wmertens committed:
nginx: add gzip_vary to recommended settings

Google PageSpeed recommends turning this on to allow proxies to cache

FPtje committed:
Strongswan after network-online instead of network

The systemd service file shipped with strongswan has strongswan started after `network-online`. It turns out that this is for good reason: failure to connect on boot otherwise. 

See this thread on the mailing list, which my colleague initiated after finding that our NixOS strongswan config wouldn't connect on boot:
https://lists.strongswan.org/pipermail/users/2017-January/010359.html

Tested on a local config (which has the strongswan service config overridden).

jtojnar committed:
corebird: fix video playback

Corebird requires gst-plugins-bad compiled with GTK support to play
videos.

This commit enables the GTK support.

fpletz committed:
imagemagick: 6.9.8-10 -> 6.9.9-0

fpletz committed:
imagemagick7: 7.0.6-0 -> 7.0.6-1

ttuegel committed:
vcsh: Patch for Git 2.13.2

Fixes #27134.

copumpkin committed:
pants: 1.2.1 -> 1.3.0

I'm also removing the pants13-pre attribute since it's unnecessary and I
don't want to maintain 1.4.x prereleases yet.

I'm temporarily taking out the pants native rust engine shenanigans until
I can get a nix-native rust build of it to work (it works on Linux but
not on Darwin!)

jwiegley committed:
coqPackages.compcert: Recent compcert supports 64-bit architectures

FRidh committed:
Merge pull request #27005 from jtojnar/fix/corebird-gst

gst_all_1.gst-plugins-bad: build gtksink plugin

FRidh committed:
Merge pull request #27445 from dapphub/dapp

dapp: init at v0.5.2

dezgeg committed:
iproute: 4.11.0 -> 4.12.0

dezgeg committed:
pv: 1.6.0 -> 1.6.6

dezgeg committed:
strace: 4.17 -> 4.18

copumpkin committed:
pants13-pre: remove from pythonPackages

Sorry, I broke evaluation in my last pants commit!

globin committed:
programs.gnupg: use extraInit instead of interactiveShellInit

Otherwise some programmes cannot use the GPG agent, e.g. applications
started from dmenu.

Behaviour was changed in #26888, this reverts that part.

Mic92 committed:
Merge pull request #27057 from Nadrieril/bitlbee-libpurple

bitlbee service: Add option to load libpurple plugins into bitlbee

peti committed:
LTS Haskell 8.23

peti committed:
hackage-packages.nix: automatic Haskell package set update

This update was generated by hackage2nix v2.3.1-12-gcb2265e from Hackage revision
https://github.com/commercialhaskell/all-cabal-hashes/commit/f69cfc36c9bc2a52f9c3de9d0cc823a00db59705.

lukateras committed:
html-xml-utils: 6.9 → 7.1, macOS support

joachifm committed:
Merge pull request #27450 from mdorman/emacs-updates

Semi-automated Emacs package updates

Mic92 committed:
Merge pull request #27454 from yegortimoshenko/patch-5

html-xml-utils: 6.9 → 7.1, macOS support

LnL7 committed:
Merge pull request #26429 from aherrmann/pr_llvm_4_fix_debug

llvm_4: Fix debugVersion

Mic92 committed:
rustRegistry: 2017-06-27 -> 2017-07-17

hectorj committed:
nomad: 0.5.5 -> 0.5.6

aristidb committed:
Merge pull request #27436 from np/keybase-update

Keybase updates

joachifm committed:
Merge pull request #27441 from jfrankenau/fix-solvespace

solvespace: Fix path in desktop entry file

gebner committed:
iprover: 0.8.1 -> 2.5

aristidb committed:
wireguard: allow not storing private keys in world-readable /nix/store (#27433)

* wireguard: allow not storing private keys in world-readable /nix/store

globin committed:
gnome3.gexiv2: 0.10.4 -> 0.10.6

peterhoeg committed:
qtox: 1.10.2 -> 1.11.0

rycee committed:
expat: 2.2.1 -> 2.2.2

billksun committed:
Add cmakeFlags for auto-type and yubikey (#27321)

* Add cmakeFlags for auto-type and yubikey

* Add libyubikey dependency

* Add auto-type dependencies

* Add new dependencies to buildInputs

* Fix duplication and capitalization in dependencies

* Add yubikey-personalization to dependency and build

* Add release build flag

* Revert "Add release build flag"

This reverts commit afabd63db0ad68348909c67b488787501b7991bb.

FRidh committed:
heroku: mark as broken

jameysharp committed:
build-fhs-userenv: Propagate $XAUTHORITY

The `DISPLAY` environment variable is propagated into chroots built with
`buildFHSUserEnv`, but currently the `XAUTHORITY` variable is not. When
the latter is set, its value is usually necessary in order to connect to
the X server identified by the former.

This matters for users running gdm3, for example, who have `XAUTHORITY`
set to something like `/run/user/1000/gdm/Xauthority` instead of the X
default of `~/.Xauthority`, which doesn't exist in that setup.

Fixes #21532.

Mic92 committed:
Merge pull request #27410 from florianjacob/journalwatch

journalwatch & journalwatch service: init at 1.1.0

Mic92 committed:
Merge pull request #27462 from jameysharp/patch-1

build-fhs-userenv: Propagate $XAUTHORITY

edolstra committed:
nix: 1.11.12 -> 1.11.13

rbvermaa committed:
Add file with GCE image locations, similar to ec2-amis.nix. Will be used by nixops.

(cherry picked from commit 9d810ddcc1938a90090fd60f8924f4e83acbeee2)

rbvermaa committed:
Add file with Azure image locations, similar to ec2-amis.nix. Will be used by nixops.

(cherry picked from commit e93f26847ea41cce6633b6a0feb6ce31b0722d5d)

Mic92 committed:
zfsUnstable: mark as stable with 4.12

Commit:
Merge pull request #27458 from ttuegel/bugfix/kde-icon-cache

breeze-icons: generate icon cache

ttuegel committed:
mdadm: unset STRIP

Commit 093cc00cdd9d8cf31ecce5bc1dd3645c460a1b98, sets the STRIP environment
variable by default, but this confuses the mdadm Makefile, which uses STRIP as a
flag to `install'.

ttuegel committed:
openssh: unset LD

Commit 093cc00cdd9d8cf31ecce5bc1dd3645c460a1b98, sets the LD environment
variable by default, but this confuses the openssh Makefile because `configure'
does not respect it.

ttuegel committed:
ppp: fix invalid use of substituteInPlace

substituteInPlace was invoked with multiple targets on the command line, which
is not supported.

ttuegel committed:
fetchurl: remove unpaired call to `stopNest'

Fixes #27406.

Commit 5d4efb2c816d2143f29cad8153faad1686557b2a added an assertion to `stopNest'
which requires it be correctly paired with `startNest'. `fetchurl' calls
`stopNest', but never calls `startNest'; the former calls are removed.

edolstra commented:
I disagree with the `CONTRIBUTOR` part. First, I've never used `nox` myself, so I'm hesistant to make it a hard requirement for other contributors. 

Second, requiring somebody to test all reverse dependencies is pretty extreme: for example, it would require that if you upgrade gcc, you have to test *every* package before even opening a PR. Making sure that reverse dependencies keep working is also the responsibility of the maintainers of those packages! (For example, I'm going to merge the GCC 5 -> 6 upgrade soon, and I'm sure not going to fix every obscure package myself...)

ttuegel committed:
newt: unset CPP

Commit 093cc00cdd9d8cf31ecce5bc1dd3645c460a1b98 sets the CPP environment
variable by default, confusing the newt Makefile, which expects CPP=gcc for
computing dependencies.

ttuegel committed:
systemd: fix broken source hash

Ericson2314 commented on pkgs/build-support/fetchurl/builder.sh:
```diff
@@ -39,7 +39,6 @@ tryDownload() {
           curlexit=$?;
        fi
     done
-    stopNest
```
Does anyone know why these were unbalanced to begin with?

ttuegel commented:
> I disagree with the CONTRIBUTOR part. First, I've never used nox myself, so I'm hesistant to make it a hard requirement for other contributors.

I have a fix in staging that has been blocked for days while I clean up someone else's mess. Sweeping changes were made to `stdenv` with--as far as I can tell--no testing. IIUC, `staging` is for building binaries in anticipation of merging into `master`; it's not a dumping ground for half-finished experiments. ATM, you can't build even a simple system off `staging`. You can't even build Nix itself on `staging`!

My radical suggestions will prevent this from happening in the future, but if you don't like them, come up with something better.

Ericson2314 commented on pkgs/tools/networking/ppp/default.nix:
```diff
@@ -34,7 +34,9 @@ stdenv.mkDerivation rec {
   postFixup = ''
-    substituteInPlace $out/bin/{pon,poff,plog} --replace "/usr/sbin" "$out/bin"
+    for tgt in pon poff plog; do
+      substituteInPlace "$out/bin/$tgt" --replace "/usr/sbin" "$out/bin"
```
Do you know what the underlying error is here (I'll find it if you don't off-hand)? I didn't intend to change the interface of substituteInPlace so I should fix that rather than this package need be changed.

ttuegel replied:
AFAIK, `substituteInPlace` has never accepted multiple targets. At least, that is not the documented behavior. I don't know how this ever worked!

Ericson2314 replied:
Whew!

FRidh commented:
Requiring to always test all reverse dependencies is indeed too extreme. But I think we all agree that `stdenv` requires more testing. The guidelines in the Nixpkgs manual already state what we expect from contributors so I suggest to extend that part instead. 

Hopefully in the future we could have CI again for pull requests. Or simply Hydra jobs for those that make (major) changes since not everyone has access to sufficiently powerful machines to build all reverse dependencies.

ttuegel commented:
> The guidelines in the Nixpkgs manual already state what we expect from contributors so I suggest to extend that part instead.

What is the difference between listing a policy in the Nixpkgs manual, listing it in `CONTRIBUTING.md`, and listing it in the PR template? Except for the more detailed instructions about testing `stdenv`, everything I added to `CONTRIBUTING` is already a checkbox in the PR template. Why do we have those boxes if we don't expect people to check them?

> Hopefully in the future we could have CI again for pull requests. Or simply Hydra jobs for those that make (major) changes since not everyone has access to sufficiently powerful machines to build all reverse dependencies.

This is the sort of thing that is nearly universal among other projects the size of Nixpkgs. Until we have a policy about this, and infrastructure to back it up, this kind of problem will continue to make working on Nixpkgs an often miserable experience. ☹️

ttuegel committed:
libunistring: disable parallel building

Parallel building causes a test deadlock and has been disabled.

ttuegel committed:
systemd: unset RANLIB

Commit 093cc00cdd9d8cf31ecce5bc1dd3645c460a1b98 sets the RANLIB environment
variable by default, causing `make' to invoke the wrong program.

FRidh commented:
> What is the difference between listing a policy in the Nixpkgs manual, listing it in CONTRIBUTING.md, and listing it in the PR template? Except for the more detailed instructions about testing stdenv, everything I added to CONTRIBUTING is already a checkbox in the PR template. Why do we have those boxes if we don't expect people to check them?

My view is that those checkboxes are to communicate what the contributor has done. It's helpful information that you cannot see from the PR contents. They normally should have done those things before opening a PR, and this is just for them to say they did or did not.

ttuegel committed:
x264: unset AS

Commit 093cc00cdd9d8cf31ecce5bc1dd3645c460a1b98 sets the environment variable
`AS' to the binutils assembler, but x264 needs yasm.

ttuegel committed:
zfs: fix invalid use of substituteInPlace

substituteInPlace was invoked with multiple targets on the command line, which
is not supported.

ttuegel committed:
grub2: unset CPP

Commit 093cc00cdd9d8cf31ecce5bc1dd3645c460a1b98 sets the environment variable
`CPP' by default, but this intereferes with dependency calculation.

Ericson2314 commented:
> This is the sort of thing that is nearly universal among other projects the size of Nixpkgs. Until we have a policy about this, and infrastructure to back it up, this kind of problem will continue to make working on Nixpkgs an often miserable experience. ☹️

Right I do think individual users do not necessarily have the compute capacity to rebuild the world, but `hydra.nixos.org` does. For non-mass rebuild PRs, this is easy. For mass-rebuild PRs, we'll have to be clever about resource allocation, but I think the problem is still solvable. CF: https://internals.rust-lang.org/t/batched-merge-rollup-feature-has-landed-on-bors/1019

edolstra commented:
> This is the sort of thing that is nearly universal among other projects the size of Nixpkgs.

What projects routinely build tens of thousands of packages whenever some user opens a PR?

ttuegel commented:
> What projects routinely build tens of thousands of packages whenever some user opens a PR?

I was referring to the size of the codebase and the number of contributors, the main features that determine the difficulty of coordinating efforts across teams of contributors and subsets of packages. But you're right: the dominant feature here is the resources required by CI. You actually make a good case against contributing to Nixpkgs in the first place: you can't save a drowning man by pouring more water on him.

ttuegel committed:
qemu: unset CPP

Commit 093cc00cdd9d8cf31ecce5bc1dd3645c460a1b98 sets the environment variable
`CPP' by default, but this interferes with dependency calculation.

ttuegel committed:
qt4: unset LD

Commit 093cc00cdd9d8cf31ecce5bc1dd3645c460a1b98 sets the `LD' environment
variable by default, interfering with the Makefile which uses gcc for linking.

ttuegel committed:
syslinux: fix invalid use of substituteInPlace

substituteInPlace was invoked with multiple targets on the command line, which
is not supported.

ttuegel committed:
Merge branch 'master' into bugfix/staging/stdenv

joachifm commented:
In the short term, can @Ericson2314 and others who wish to work on fundamental features be allotted a jobset(s) on hydra, so that `staging` can return to serve its original purpose? I know there are some such jobsets already but it's not at all transparent how they are allocated.

knedlsepp commented on pkgs/development/libraries/libunistring/default.nix:
```diff
@@ -14,6 +14,8 @@ stdenv.mkDerivation rec {
 
   propagatedBuildInputs = stdenv.lib.optional (!stdenv.isLinux) libiconv;
 
+  enableParallelBuilding = false;
```
I've been having trouble building `libunistring` on openSUSE using nix, which I only solved by using an overlay that disabled tests for that particular library. It always got stuck in the `test-lock` test with full CPU usage. Does this solve that particular issue? I think I had the same problem with `findutils`. Let me see if I can recreate the issue on that particular machine.

ttuegel committed:
nixos/tests/keymap: use SLIM theme from nixos/tests/slim

ttuegel committed:
jam: unset AR

Commit 093cc00cdd9d8cf31ecce5bc1dd3645c460a1b98 sets the AR environment variable
by default, but this causes the jam Makefile to use the wrong command.

ttuegel committed:
argyllcms: unset AR

Commit 093cc00cdd9d8cf31ecce5bc1dd3645c460a1b98 sets the AR environment variable
by default, but this causes the argyllcms Makefile to use the wrong command.

ttuegel commented:
At this stage, all the NixOS tests are passing which were not already broken before the stdenv update, so I am going to merge this.

ttuegel commented:
@edolstra Hydra hasn't evaluated the `staging` branch. Could you nudge it in the right direction?

FRidh commented:
@ttuegel http://hydra.nixos.org/eval/1377845 includes these changes

Ericson2314 commented:
@joachifm I tried to use http://hydra.nixos.org/jobset/nixpkgs/ericson2314-cross-staging, but it never would evaluate. :/

Ericson2314 commented:
@ttuegel Wow. Thank you so much! I really owe you for cleaning up my mess---collect that debt someday.

domenkozar commented:
@ttuegel thank you man :)

domenkozar commented:
Go seems to be still broken: `/nix/store/xilxav9i6xkrra3vclmlm7zwbbvr46qv-go-1.8.3/share/go/pkg/tool/linux_amd64/link: too many errors`  at http://hydra.nixos.org/build/56796101

@Ericson2314 any clues?

FRidh commented:
@domenkozar @Ericson2314 issue for Go https://github.com/NixOS/nixpkgs/issues/27589

copumpkin commented:
I second giving @Ericson2314 his own Hydra job for testing this. When I was making sweeping changes to the Darwin stdenv (which is if anything less impactful to everyone than the work he's doing), I used the `copumpkin-wip` job on Hydra and it was extremely helpful for finding things I would've never found otherwise. Combined with "incrementalism" in my changes, that allowed me to make some pretty sweeping changes to a lot of the Darwin fundamentals a few months ago, without anyone really noticing (at least not negatively).

globin commented:
@edolstra @rbvermaa ^

Ericson2314 commented:
I have hydra perms to make jobs, and @grahamc realized the problem was simply that I had previous disabled my job 😊. We're all good.
